### PR TITLE
New feature : Email / Slack / Telegram notifications

### PR DESCRIPTION
**Pull request by laxdog**
Extra module to allow notifications across messaging platforms

## Description
Currently this only supports the 3 mentioned, but should be fairly easy to extend. There's one central method that will push a message to all configured platforms when you call log.notify()

I still need to add some of the hooks, summary every X minutes and when a loan is made over the xdaythreshold. 

I'm actually wondering about the best way to do the summary is. I was initially going to use the schedule module, but since I bothered to add all these in pure python I thought I may as well try to stay away from external modules. I've had a quick look at sched, which seems it might work. 

I still have a few TODO items, including all the docs and fixing up a few things I left out.

This closes https://github.com/Mikadily/poloniexlendingbot/issues/153

## TESTING STAGE
Currently running

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] **I have read CONTRIBUTING.md**
- [x] **I fully understand [Github Flow.](https://guides.github.com/introduction/flow/)**
- [x] **My code adheres to the [code style of this project.](https://poloniexlendingbot.readthedocs.io/en/latest/contributing.html)**
- [x] **I have updated the documentation in /docs if I have changed the config, arguments, logic in how the bot works, or anything that understandably needs a documentation change.**
- [x] **I have updated the config file accordingly if my change requires a new configuration setting or changes an existing one.**
- [x] **I have tested the bot with no issues for 24 continuous hours. If issues were experienced, they have been patched and tested again.**

### CODE REVIEW

File: default.cfg.example [hunk 1]
```diff
@@ -125,3 +125,24 @@ hideCoins = True
...
 #maxpercenttolend = 0
 #maxtolendrate = 0
+
+[Notifications]
+notify_tx_coins = True
+notify_xday_threshold = True
+notify_summary_minutes = 60
+notify_caught_exception = True
+
+email = True
```
Comment (Evanito) [inline default.cfg.example]: Should probably set these to `False` by default, in case someone wants to do minimal configuring.

Comment (rnevet) [inline default.cfg.example]: I agree, notifications should be by default off.

**Comment by rnevet** [comment]
Need to update documentation.

**Comment by laxdog** [comment]
Yea, lots of doc updates to do. Hoping to get some time over the next week to close out the final modifications. I sat down to do them and ended up doing that other pull request instead.

**Comment by laxdog** [comment]
I think this is done from my side, pending any modifications.

**Comment by laxdog** [comment]
I don't really agree with all those Codacy issues, but I'll change them if you want to keep it consistent.

**Comment by rnevet** [comment]
@laxdog isn't that related to Python 2 vs 3?

**Comment by laxdog** [comment]
Yea, you don't have to use them in 2, but you can if you want. You have to use them in 3 as it's a function. Which is why I use them, it's backwards compatible. 

I only said I'd change it if you want, as the rest of the project doesn't use them.

**Comment by rnevet** [comment]
@Evanito can you re-review?

**Comment by Evanito** [comment]
Just removed the parenthesis check from Codacy. I have been swamped the past weeks, glad to have some free time.

File: docs/configuration.rst [hunk 2]
```diff
@@ -311,4 +311,54 @@ You can pass options to statistics page by adding them to URL. Eg, ``http://loca
 
 ``earningsInOutputCurrency`` define which earnings are show in output currency
 - Allowed values: ``all``, ``summary``
-- Default value: ``all``
\ No newline at end of file
+- Default value: ``all``
+<<<<<<< HEAD
```
Comment (Evanito) [inline docs/configuration.rst]: Byproduct of fixing a merge conflict, please remove.
Lines 315 to 320

File: docs/configuration.rst [hunk 3]
```diff
@@ -311,4 +311,54 @@ You can pass options to statistics page by adding them to URL. Eg, ``http://loca
...
+The bot supports sending notifications for serveral different events on several different platforms. To enable notifications, you must first have a section in your config called ``[notifications]``, inside which you should enable at least one of the following events and also at least one notification platfom. The list of events you can notify about are:
+
+notify_tx_coins
+    This will send a notification if any coins are transferred from your exchange account, to your lending account. You must have ``transferableCurrencies`` enabled for this to work.  Then you should set ``notify_tx_coins = True``.
+
+notify_xday_threshold
+    This will send a notifcation every time a loan is created that is above your ``xdaythreshold`` config value. To enable you should set ``notify_xday_threshold = True``.
+
+notify_summary_minutes
+    This will send a summary of the current loans you have every X minutes. This is similar to the information you get in the log line when running the bot, or the line a the top of the web page. To enable this add ``notify_summary_minutes = 120``. This will send you a notifcation every 2 hours (120 minutes)
```
Comment (Evanito) [inline docs/configuration.rst]: typo "notifcation"

File: modules/Configuration.py [hunk 4]
```diff
@@ -121,3 +125,34 @@ def get_currencies_list(option):
         return list(set(cur_list))
     else:
         return []
+
+
+def get_notification_config():
+    notify_conf = {}
+    notify_conf['enable_notifications'] = config.has_section('Notifications')
```
Comment (Evanito) [inline modules/Configuration.py]: Case sensitive section name?

File: modules/Notify.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
+# coding=utf-8
+import urllib
+import urllib2
+import json
+import smtplib
+
+# TODO
```
Comment (Evanito) [inline modules/Notify.py]: Remove TODO if done

File: modules/Notify.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+def post_to_telegram(msg, chat_ids, bot_id):
+    for chat_id in chat_ids:
+        post_data = {"chat_id": chat_id, "text": msg}
+        url = "https://api.telegram.org/bot" + bot_id + "/sendMessage"
+        try:
+            response = urllib2.urlopen(url, urllib.urlencode(post_data))
+            check_urlib_response(response, 'telegram')
+        except urllib2.HTTPError as e:
+            msg = "Your bot id is probably configured incorrectly"
+            raise NotifcationException("{0}\n{1}".format(e, msg))
```
Comment (Evanito) [inline modules/Notify.py]: typo

File: modules/Notify.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+    try:
+        server = smtplib.SMTP_SSL(email_smtp_server, email_smtp_port)
+        server.ehlo()
+        server.login(email_login_address, email_login_password)
+        server.sendmail(email_login_address, email_to_addresses, email_text)
+        server.close()
+    except Exception as e:
+        print("Cound not send email, got error {0}".format(e))
+        raise NotifcationException(e)
```
Comment (Evanito) [inline modules/Notify.py]: typo

File: modules/Notify.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        if isinstance(v, unicode):
+            v = v.encode('utf8')
+        elif isinstance(v, str):
+            # Must be encoded in UTF-8
+            v.decode('utf8')
+        out_dict[k] = v
+    return out_dict
+
+
+class NotifcationException(Exception):
```
Comment (Evanito) [inline modules/Notify.py]: Notif-i-cation

**Comment by laxdog** [comment]
Sorry for the number of commits, I had to use the online editor.

**Comment by rnevet** [comment]
Crazy!! we have notifications!

**Comment by laxdog** [comment]
Really happy to get this in btw, I've been using it a lot. It's really useful for me. The notification on exception and transfer from exchange account are particularly useful.

Thanks for merging!

**Comment by rnevet** [comment]
@laxdog Thanks for for taking the initiative, your help is appreciated.

**Comment by praeluceo** [comment]
That is awesome laxdog, the docs need a little more work, or I'm doing something wrong. I'll file a bug if I can't figure it out, but I think the issue is that with telegram, "telegram_bot_id" should be renamed to "telegram_bot_token", and "telegram_chat_ids" should be broken up into "telegram_bot_name" and "telegram_bot_username", but that's at least how I'm trying to use it, but I get an error with each invocation, despite changing the positions of variables:


```
Welcome to Poloniex Lending Bot
Started WebServer, lendingbot status available at http://127.0.1.1:8000/lendingbot.html, http://127.0.0.1:8000/lendingbot.html
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/lib/python2.7/sched.py", line 117, in run
    action(*argument)
  File "/home/ronald/bots/poloniexlendingbot/modules/Lending.py", line 96, in notify_summary
    log.notify(Data.stringify_total_lended(*Data.get_total_lended()), notify_conf)
  File "/home/ronald/bots/poloniexlendingbot/modules/Logger.py", line 150, in notify
    send_notification(msg, notify_conf)
  File "/home/ronald/bots/poloniexlendingbot/modules/Notify.py", line 83, in send_notification
    post_to_telegram(msg, nc['telegram_chat_ids'], nc['telegram_bot_id'])
  File "/home/ronald/bots/poloniexlendingbot/modules/Notify.py", line 50, in post_to_telegram
    raise NotificationException("{0}\n{1}".format(e, msg))
NotificationException: HTTP Error 403: Forbidden
Your bot id is probably configured incorrectly

```

**Comment by laxdog** [comment]
OK, sorry. I'd never used telegram before this implementation, so I'm probably off on the terms. 

I've a few things to fix next week, so I'll fix this then. Thanks for the feedback, it's really useful. 

I also see you have the same address twice in your bot web address. I'll fix that too.

**Comment by praeluceo** [comment]
Something I noticed, was what looked like incomplete variables, take a look at my diff below (it doesn't work still, but it did change the error message, and I should probably just file a bug on it):

```
diff --git a/modules/Notify.py b/modules/Notify.py
index 1acadf2..795f058 100644
--- a/modules/Notify.py
+++ b/modules/Notify.py
@@ -38,10 +38,10 @@ def post_to_slack(msg, channels, token):
         check_urlib_response(response, 'slack')
 
 
-def post_to_telegram(msg, chat_ids, bot_id):
-    for chat_id in chat_ids:
+def post_to_telegram(msg, telegram_chat_ids, telegram_bot_id):
+    for chat_id in telegram_chat_ids:
         post_data = {"chat_id": chat_id, "text": msg}
-        url = "https://api.telegram.org/bot" + bot_id + "/sendMessage"
+        url = "https://api.telegram.org/bot" + telegram_bot_id + "/sendMessage"
         try:
             response = urllib2.urlopen(url, urllib.urlencode(post_data))
             check_urlib_response(response, 'telegram')

```

**Comment by Fabien98** [comment]
Hello,

I'm having the same issue here:

```
Welcome to Poloniex Lending Bot
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/home/Fabien/poloniexlendingbot/modules/MarketAnalysis.py", line 54, in update_market_loop
    delete_old_data()
  File "/home/Fabien/poloniexlendingbot/modules/MarketAnalysis.py", line 74, in delete_old_data
    if get_day_difference(row[0]) < max_age:
IndexError: list index out of range


WARN: Exception found when analysing markets, if this happens for more than a couple minutes please make a Github issue so we can fix it. Otherwise, you can safely ignore it. Error: list index out of range
Exception in thread Thread-2:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/lib/python2.7/sched.py", line 117, in run
    action(*argument)
  File "/home/Fabien/poloniexlendingbot/modules/Lending.py", line 97, in notify_summary
    log.notify(Data.stringify_total_lended(*Data.get_total_lended()), notify_conf)
  File "/home/Fabien/poloniexlendingbot/modules/Logger.py", line 150, in notify
    send_notification(msg, notify_conf)
  File "/home/Fabien/poloniexlendingbot/modules/Notify.py", line 83, in send_notification
    post_to_telegram(msg, nc['telegram_chat_ids'], nc['telegram_bot_id'])
  File "/home/Fabien/poloniexlendingbot/modules/Notify.py", line 50, in post_to_telegram
    raise NotificationException("{0}\n{1}".format(e, msg))
NotificationException: HTTP Error 400: Bad Request
Your bot id is probably configured incorrectly
```
My config setting is similar to @praeluceo 

Should we put the name of the channel somwhere as well?

**Comment by racydex** [comment]
Hi Guys!
I'm new here. Hope this isn't the wrong place to post this error.
Trying to get Telegram notifications to work but getting this error. Bot id is configured correctly as I tested on a browser and it works fine. Can someone explain what the 400 error is?
```
Error during summary notification: HTTP Error 400: Bad Request
Your bot id is probably configured incorrectly
```

Apologies again if this ain't the right place to post this error.
Appreciate any assistance. I'm not a coder.

**Comment by laxdog** [comment]
Hey @Fabien98 and @racydex , I did a whole update of the docs, after this issue was reported. It explains everything as a few people were having problems. You can see the Telegram section here: http://poloniexlendingbot.readthedocs.io/en/latest/configuration.html#notifications

**Comment by Fabien98** [comment]
@laxdog Thanks for this update.
I saw the new docs and managed to get  the telegram notifications thanks to this easy step-by-step.
Thank you!

**Pull request merged**